core/src/main/scala/org/apache/spark/scheduler/JobWaiter.scala

```diff
@@ -28,12 +31,18 @@ private[spark] class JobWaiter[T](
     resultHandler: (Int, T) => Unit)
   extends JobListener {
 
+  private val promise = Promise[Unit]
+
   private var finishedTasks = 0
 
   // Is the job as a whole finished (succeeded or failed)?
   @volatile
   private var _jobFinished = totalTasks == 0
 
+  if (_jobFinished) {
```

Comment:
@zsxwing, this `if` statement fixes a subtle bug that we found in your JobWaiter future: if a job has no tasks, then it is marked as finished immediately and `taskSucceeded` will not be called, so we need to complete the promise here.  We noticed this because a test in AsyncRDDActionsSuite was hanging.